common.blocks/i-bem-dom/__event-manager/_type/i-bem-dom__event-manager_type_bem.js

```diff
@@ -0,0 +1,231 @@
+    ],
+    function(
+        provide,
+        BEMINTERNAL,
+        inherit,
+        identify,
+        objects,
+        $) {
+
+var undef,
+    EVENT_PREFIX = '__bem__',
+
+    eventStorage = {},
+
+    buildEventName = function(e, entityName) {
+        return EVENT_PREFIX + entityName + (typeof e === 'object'? e.type : e);
+    },
+
+    buildBemEventManagerParams = function(ctx, defCtx, defSelector, defCls, getEntityCls) {
+        var res = {
+                entityCls : null,
+                ctx : defCtx,
+                selector : defSelector,
+                key : ''
+            };
+
+        if(ctx) {
+            var typeOfCtx = typeof ctx;
+
+            if(typeOfCtx === 'object' && ctx.__self) { // bem entity instance
+                res.ctx = ctx.domElem;
+                res.key = ctx._uniqId;
+                res.entityCls = ctx.__self;
+            } else if(typeOfCtx === 'string' || typeOfCtx === 'object' || typeOfCtx === 'function') {
+                var entityName, blockName, elemName, modName, modVal;
+                if(typeOfCtx === 'string') { // elem name
+                    blockName = defCls._blockName;
+                    elemName = ctx;
+                } else if(typeOfCtx === 'object') { // bem entity with optional mod val
+                    blockName = ctx.block?
+                        ctx.block.getName() :
+                        defCls._blockName;
+                    elemName = typeof ctx.elem === 'function'?
+                        ctx.elem.getName() :
+                        ctx.elem;
+                    modName = ctx.modName;
+                    modVal = ctx.modVal;
+                } else if(ctx.getName() === ctx.getEntityName()) { // block class
+                    blockName = ctx.getName();
+                } else { // elem class
+                    blockName = defCls._blockName;
+                    elemName = ctx.getName();
+                }
+
+                entityName = BEMINTERNAL.buildClass(blockName, elemName);
+                res.entityCls = getEntityCls(entityName);
+                res.selector = '.' + (res.key = entityName + BEMINTERNAL.buildModPostfix(modName, modVal));
+            }
+        } else {
+            res.entityCls = defCls;
+        }
+
+        return res;
+    },
+
+    EventManager = inherit({
+        __constructor : function(params, fnWrapper) {
+            this._params = params;
+            this._fnWrapper = fnWrapper;
+            this._storage = {};
+        },
+
+        /**
+         * Adds an event handler
+         * @param {String} e Event type
+         * @param {Object} [data] Additional data that the handler gets as e.data
+         * @param {Function} fn Handler
+         * @returns {EventManager} this
+         */
+        on : function(e, data, fn) {
+            if(typeof e === 'object') {
+                objects.each(e, function(fn, e) {
+                    this.on(e, data, fn);
+                }, this);
+            } else {
```

Comment:
I'm not sure why you avoid early returns but code would look much simpler with them.

If we `return this` if that if we could move out `else` clause and "untab" the rest of function.